py/farm_ng/core/event_service_recorder.py

```diff
@@ -0,0 +1,362 @@
+        """Stops recording the events to a file."""
+        # do nothing if not recording
+        if self._recorder is None:
+            return
+
+        self._recorder.logger.info("stopping recording")
+
+        # cancel the task and set the recorder to None
+        self._recorder_task.cancel()
+        self._recorder = None
+        self._recorder_task = None
+
+    # private methods
+
+    async def _request_reply_handler(
+        self, event_service: EventServiceGrpc, request: RequestReplyRequest
+    ) -> Message:
+        """Handles the request reply.
+
+        Args:
+            event_service (EventServiceGrpc): the event service.
+            request (RequestReplyRequest): the request.
+
+        Returns:
+            Message: the response message.
+        """
+
+        if request.event.uri.path == "start":
+            config_list: EventServiceConfigList = payload_to_protobuf(
+                request.event, request.payload
+            )
+            event_service.logger.info("start: %s", config_list)
+            file_base = self._data_dir.joinpath(get_file_name_base())
+            await self.start_recording(file_base, config_list)
+            return StringValue(value=str(file_base))
+        elif request.event.uri.path == "stop":
+            await self.stop_recording()
+        return Empty()
+
+
+def service_command(_args):
+    config_list, service_config = load_service_config(args)
+    event_service: EventServiceGrpc = EventServiceGrpc(
+        grpc.aio.server(), service_config
+    )
+    recorder_service = RecorderService(event_service)
+
+    async def job():
+        # define the async tasks
+        async_tasks: list[asyncio.Task] = []
+
+        # add the service task
+        async_tasks.append(event_service.serve())
+
+        # run the tasks
+        await asyncio.gather(*async_tasks)
+
+    asyncio.get_event_loop().run_until_complete(job())
+
+
+def client_start_command(_args):
+    config_list, service_config = load_service_config(args)
+    loop = asyncio.get_event_loop()
+
+    async def job():
+        reply = await EventClient(service_config).request_reply("start", config_list)
+        print(payload_to_protobuf(reply.event, reply.payload))
+
+    loop.run_until_complete(job())
+
+
+def client_stop_command(_args):
+    config_list, service_config = load_service_config(args)
+    loop = asyncio.get_event_loop()
+    loop.run_until_complete(EventClient(service_config).request_reply("stop", Empty()))
+
+
+def record_command(_args):
+    async def job():
+        configs = proto_from_json_file(_args.service_config, EventServiceConfigList())
+        recorder = EventServiceRecorder(_args.config_name, config_list=configs)
+        await recorder.subscribe_and_record(
+            _args.output, _args.extension, _args.max_file_mb
+        )
+
+    asyncio.get_event_loop().run_until_complete(job())
+
+
+if __name__ == "__main__":
+    parser = argparse.ArgumentParser()
```

Comment:
not in favor of having both the library definition and the code using it in the same place in order to keep clean of the tools needed by the library and what is needed to run it. For this reason they added the ability to define entrypoints in the `setup.py`. See: https://setuptools.pypa.io/en/latest/userguide/entry_point.html

Reply:
i'd suggest to use some of the well supported community packages like `Typer` which has all the bells and whistles for building modern cli tools. See: https://typer.tiangolo.com/

Reply:
I'm not in favor of a new dependency for argument parsing, like typer.  The goal here is to have a simple uniform command line interface for all services, and argparse is built in to python.

python -m farm_ng.core.event_service_recorder --service-config config.json --service-name name_of_my_service

This allows for generic launching code with no specific knowledge of the implementation details of the service.

As far as entry points are concerned I'm not sure I understand the proposal, in this case it would be:

```
[project.scripts]
farm-ng-core-event-service-recorder = "farm_ng.core:event_service_recorder"
```

Reply:
going to resolve this for now.